libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1411 @@
+{
+  return filter_flatpak_ref_actions_table (ref_actions,
+                                           keep_only_existing_actions,
+                                           progresses);
+}
+
+const gchar *
+eos_updater_util_pending_flatpak_deployments_state_path (void)
+{
+  return eos_updater_get_envvar_or ("EOS_UPDATER_TEST_UPDATER_FLATPAK_UPGRADE_STATE_DIR",
+                                    LOCALSTATEDIR "/lib/eos-application-tools/flatpak-autoinstall.progress");
+}
+
+const gchar *
+eos_updater_util_flatpak_autoinstall_override_paths (void)
+{
+  return eos_updater_get_envvar_or ("EOS_UPDATER_TEST_UPDATER_FLATPAK_AUTOINSTALL_OVERRIDE_DIRS",
+                                    LOCALSTATEDIR "/lib/eos-application-tools/flatpak-autoinstall.d;"
+                                    SYSCONFDIR "/eos-application-tools/flatpak-autoinstall.d");
+}
+
+GHashTable *
+eos_updater_util_flatpak_ref_action_application_progress_in_state_path (GCancellable  *cancellable,
+                                                                        GError       **error)
+{
+  const gchar *state_file_path = eos_updater_util_pending_flatpak_deployments_state_path ();
+  g_autoptr(GKeyFile) state_key_file = g_key_file_new ();
+  g_auto(GStrv) groups = NULL;
+  g_autoptr(GHashTable) ref_action_progress_for_files = g_hash_table_new_full (g_str_hash,
+                                                                               g_str_equal,
+                                                                               g_free,
+                                                                               NULL);
+
+  GStrv groups_iter = NULL;
+  g_autoptr(GError) local_error = NULL;
+
+  /* Read the key file for sections about the application progress
+   * of each autoinstall file */
+  if (!g_key_file_load_from_file (state_key_file,
+                                  state_file_path,
+                                  G_KEY_FILE_NONE,
+                                  &local_error))
+    {
+      if (!g_error_matches (local_error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return NULL;
+        }
+
+      return g_steal_pointer (&ref_action_progress_for_files);
+    }
+
+  /* Enumerate each section. The section name is the path to the file */
+  groups_iter = groups = g_key_file_get_groups (state_key_file, NULL);
+
+  while (*groups_iter != NULL)
+    {
+      const gchar *source_path = *groups_iter;
+      g_autofree gchar *progress_string = NULL;
+      guint64 progress;
+
+      /* We need to use g_key_file_get_value here to guard against errors */
+      progress_string = g_key_file_get_value (state_key_file,
+                                              *groups_iter,
+                                              "Progress",
+                                              error);
+
+      if (progress_string == NULL)
+        return NULL;
```

Comment:
If `progress_string == NULL`, should we not just not insert an entry for `source_path` in the hash table, and carry on? I think it would be fine (if unexpected) for the progress file to have an empty section — just treat it like the section doesn’t exist.

Reply:
I think this is meant to be an invariant - if there is a section in the keyfile for that filename, then it should have a `Progress` key.

Reply:
If you want. I generally put invariants into user-facing code like this only if it makes things simpler for me, and I’m not sure that’s the case here.

As long as whatever you do is documented (brief code comment is fine) I’ll be happy.